Create MTH_PS3_Full.xml

alain355 opened the pull request:
Hello
Using the 2 existing MTH PS3 definitions done by Michael Mosher (files MTH.xml and MTH_expanded.xml), I created a 3rd one, merging the CVs definitions from both files. I added a reset pane with reset comments (depending if done in POM or in Programming track mode).

AFAIK tested OK on my ESU ECoS using latest ESU SW release 4.2.2, JMRI 4.11.2 and an HO scale SD70Ace from MTH.

alain355 commented:
If someone can help on this failure, thanks.

dheap commented:
@alain355 The way you have done the Resets can be improved. I [JMRI V4.2.1](http://jmri.org/releasenotes/jmri4.2.1.shtml) I improved the `<resets>` procedure to allow specification of the applicable mode(s) for each reset, but it appears I haven't documented it anywhere.

I'll make the changes and push to this PR so you can see how it works.

alain355 commented:
Thanks Dave

dheap commented:
@alain355 I've submitted a Pull Request  [Specify reset modes in MTH_PS3_Full.](https://github.com/alain355/JMRI/pull/1) against your own JMRI patch-1 branch so you can check it before accepting it into this PR.

alain355 commented:
Hi Dave, 
I'm ready to accept your update but when I add your part to my decoder defintion (after removing my reset pane), the file isn't validated by JMRI (see below the copy of the error message when validating 1screen adding the full addon). I also tried to add each reset one by one, the 1st reset was OK, but error came when adding the 2nd one (2nd screen copy).

Before going forward, let me explain how I made the initial file (ie my issue, then this PR):
**File creation:** I used file MTH_expanded.xml as my base to create MTH_PS3_full.xml by adding the missing CVs 49, 50 and 55. Before I was obliged for my single MTH engine to create 2 definitions using the 2 existing files, depending on the Cvs I wanted to look at.
Regarding the reset part, depending on the command station and when MTH made the decoders, you can either do it in Command Mode (one reset mode only) or in Program Mode (4 modes depending on what to reset). Of course only one way to do it works. Michael Mosher, who created both previous files, confirmed (on the JMRI thread) that some older MTH PS3 loks need the file MTH.xml while other needs the 2nd one (where CVs are missing), so my file doesn't replace the 2 existing files, even if it could replace the MTH_expanded.xml one.

Part 2: trying to add you reset code
Looking at your modified reset definition, if I replaced my one with yours, my xml editor (xml marker free edition) showed me errors: it seems a </factReset> is missing after line 326. Thus I added it and tried to validate the xml from JMRI and it doesn't validate:
![image](https://user-images.githubusercontent.com/18462439/36060164-2c7bda3c-0e44-11e8-8def-96ccf47977f8.png)
So I tried to add it part by part:
If adding only the 1st "FactReset (4 lines), OK
If then adding the 2nd One NOK: 
![image](https://user-images.githubusercontent.com/18462439/36060329-ccb7461e-0e47-11e8-9817-0b7e28102158.png)

So I'm stuck there with your addition, while my older way to do it is validated.

Can you help me there?

AlanUS commented:
@dheap 
>  In JMRI V4.2.1 I improved the `<resets>` procedure to allow specification of the applicable mode(s) for each reset, but it appears I haven't documented it anywhere.

A bit of documentation would indeed be welcome; I did not notice this change when you introduced it.

dheap commented:
> @AlanUS A bit of documentation would indeed be welcome; I did not notice this change when you introduced it.

The problem is that there is absolutely no documentation of `<resets>` anywhere. I had to learn by studying what others had done and examining the schema.

It needs a whole new section in either [Creating A Custom Decoder File](http://jmri.org/help/en/html/apps/DecoderPro/CreateDecoder.shtml) or [Advanced Decoder Definitions](http://jmri.org/help/en/html/apps/DecoderPro/CreateDecoderAdvanced.shtml) and I haven't had time to do that.

alain355 commented:
As validation of my updated file, using the reset from the menu, and wo the reset pane, is OK:

- Do I need to do something more to include it in the next release?

- Shall I close it ?

Thanks for your answer Dave or Alain LM

dheap commented:
@alain355 **_Do not close this_**.  We will lose it all.

 Go to [this link](https://github.com/alain355/JMRI/pull/1). You will see a "Merge Pull Request" button on the screen Please click on it to merge it.

dheap commented:
@alain355 The link you need to click on is [https://github.com/alain355/JMRI/pull/1](https://github.com/alain355/JMRI/pull/1) Once you have done that, you will find the "Merge pull request" button you need to click.

I need to go to bed soon...

alain355 commented:
Dave, just done, enjoy your sleep (here it's lunch time). And thanks a lot.

Alain

dheap commented:
@alain355  Thanks for this. I'm off to bed now so one of the other maintainers can merge this for you after the CI tests complete. Or i'll do it tomorrow.

Just don't push anything new to your patch-1 branch or delete it until this is merged.

alain355 commented:
Have a nice sleep, Don't worry, I won't touch anything.
 Alain & Francine Carasso
acarasso_fr@yahoo.fr

      De : Dave Heap <notifications@github.com>
 À : JMRI/JMRI <JMRI@noreply.github.com> 
Cc : alain Carasso <acarasso_fr@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Samedi 10 février 2018 13h18
 Objet : Re: [JMRI/JMRI] Create MTH_PS3_Full.xml (#4876)
   
@alain355 Thanks for this. I'm off to bed now so one of the other maintainers can merge this for you after the CI tests complete. Or i'll do it tomorrow.Just don't push anything new to your patch-1 branch or delete it until this is merged.—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

dheap commented:
@alain355 Thanks. All done now and will be in the next release (and next developer build).

alain355 commented:
Thanks Dave for your helphave a nice day.
 Alain & Francine Carasso
acarasso_fr@yahoo.fr

      De : Dave Heap <notifications@github.com>
 À : JMRI/JMRI <JMRI@noreply.github.com> 
Cc : alain Carasso <acarasso_fr@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Samedi 10 février 2018 20h37
 Objet : Re: [JMRI/JMRI] Create MTH_PS3_Full.xml (#4876)
   
@alain355 Thanks. All done now and will be in the next release (and next developer build).—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

alain355 commented:
Hi DaveI downloaded the resulting file from GitHub and tried to validate it with JMRI 4.11.2. It gave an error

There's a small mistake in the italian traduction made, giving an error when validating the resulting file.

Repository file:       <mode>OPSBYTEMODE</mode>
        <label xml:lang="it">Reset delle CV salvo Tabella VelocitÃ  ai valori di fabbrica</label>
      </factReset>

Corrected file:        <mode>OPSBYTEMODE</mode>
        <label xml:lang="it">Reset delle CV salvo Tabella Velocita  ai valori di fabbrica</label>
      </factReset> Can you change the character in the repository, or shall I create a new PR?
Sorry for the trouble

Alain & Francine Carasso
acarasso_fr@yahoo.fr

      De : Dave Heap <notifications@github.com>
 À : JMRI/JMRI <JMRI@noreply.github.com> 
Cc : alain Carasso <acarasso_fr@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Samedi 10 février 2018 20h37
 Objet : Re: [JMRI/JMRI] Create MTH_PS3_Full.xml (#4876)
   
@alain355 Thanks. All done now and will be in the next release (and next developer build).—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

alain355 commented:
I forgot to add that the error is in your file Specify reset modes in MTH_PS3_Full. #1 in line 317-319
![image](https://user-images.githubusercontent.com/18462439/36071959-a552dd42-0f17-11e8-8c0e-f0f288d90dd5.png)

dheap commented:
@alain355 Interesting. It passed my local validation and CI tests, our pre-merge CI tests and the CI builds at http://jmri.tagadab.com/jenkins/job/Development/job/Builds/ where it built fine.

But I guess it could be due to your running the CI tests in a non-English language???

I also would have to raise a PR (our policy is to not directly modify the source code), so better you do it and then I/we can approve it.

alain355 commented:
I was running the PR in English (I didn't knew it was possible to do it in another language)
I've just done it Dave PR#2 Alain & Francine Carasso
acarasso_fr@yahoo.fr

      De : Dave Heap <notifications@github.com>
 À : JMRI/JMRI <JMRI@noreply.github.com> 
Cc : alain Carasso <acarasso_fr@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Dimanche 11 février 2018 11h44
 Objet : Re: [JMRI/JMRI] Create MTH_PS3_Full.xml (#4876)
   
@alain355 Interesting. It passed my local validation and CI tests, our pre-merge CI tests and the CI builds at http://jmri.tagadab.com/jenkins/job/Development/job/Builds/ where it built fine.But I guess it could be due to your running the CI tests in a non-English language???I also would have to raise a PR (our policy is to not directly modify the source code), so better you do it and then I/we can approve it.—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

alain355 commented:
Dave one question: my  PR was validated OK on my branch, thus I pushed the Merge button, anything else to do ? Alain & Francine Carasso
acarasso_fr@yahoo.fr

      De : Dave Heap <notifications@github.com>
 À : JMRI/JMRI <JMRI@noreply.github.com> 
Cc : alain Carasso <acarasso_fr@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Dimanche 11 février 2018 11h44
 Objet : Re: [JMRI/JMRI] Create MTH_PS3_Full.xml (#4876)
   
@alain355 Interesting. It passed my local validation and CI tests, our pre-merge CI tests and the CI builds at http://jmri.tagadab.com/jenkins/job/Development/job/Builds/ where it built fine.But I guess it could be due to your running the CI tests in a non-English language???I also would have to raise a PR (our policy is to not directly modify the source code), so better you do it and then I/we can approve it.—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

alain355 commented:
It was merged after the local validation, but failed  :(
Alain

      De : Dave Heap <notifications@github.com>
 À : JMRI/JMRI <JMRI@noreply.github.com> 
Cc : alain Carasso <acarasso_fr@yahoo.fr>; Mention <mention@noreply.github.com>
 Envoyé le : Dimanche 11 février 2018 11h44
 Objet : Re: [JMRI/JMRI] Create MTH_PS3_Full.xml (#4876)
   
@alain355 Interesting. It passed my local validation and CI tests, our pre-merge CI tests and the CI builds at http://jmri.tagadab.com/jenkins/job/Development/job/Builds/ where it built fine.But I guess it could be due to your running the CI tests in a non-English language???I also would have to raise a PR (our policy is to not directly modify the source code), so better you do it and then I/we can approve it.—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

dheap commented:
@alain355 What you did in merging was correct.

Let's just move on and create a new PR for the spelling/character error. Over to you to do that.

alain355 commented:
That's exactly what I've done yesterday. as I could modify you xml file in Specify reset modes in MTH_PS3_Full. #1 to replace the bad character (line 319 here https://github.com/alain355/JMRI/pull/1/files ),
I downloaded the updated decoder definition file from Github (here https://github.com/JMRI/JMRI/blob/master/xml/decoders/MTH_PS3_Full.xml), replaced in this file the bad characters, push my updated file in my created PR, submitted it, it went OK thru the 2 Travis CI tests, see below
![image](https://user-images.githubusercontent.com/18462439/36086154-e5594f6c-0fca-11e8-8898-8af5c98c6284.png)
Then I was able to ask to merge to the main branch but it crashed at the Main Travis CI in the main branch, see below
![image](https://user-images.githubusercontent.com/18462439/36086186-0b967dd0-0fcb-11e8-89ed-ac65c6d043e4.png)

I will within minutes redo it, this time I forked the file from the repository to my fork, will modify directly from the web, create a new PR, submit....
We will see the result.

alain355 commented:
The new PR  #4887 is underway.

alain355 commented:
I found another bad character new PR #4888 with the 3 characters being replaced

alain355 commented:
#4888 validated OK, can you please merge that one #4888, and stop #4887 as all updates are in #4888?

Thanks

alain355 commented:
#4888 crashed at the Travis CI after nearly 50 mn, any ideas ?
End of log below 
![image](https://user-images.githubusercontent.com/18462439/36087836-b08121b6-0fd4-11e8-9510-fb72b751a4c9.png)
